src/twisted/protocols/test/test_tls.py

```diff
@@ -1899,108 +1922,111 @@ def dataReceived(self, data):
 
         serverContextFactory = ServerTLSContext()
         wrapperFactory = TLSMemoryBIOFactory(serverContextFactory, False, serverFactory)
         sslServerProtocol = wrapperFactory.buildProtocol(None)
 
         loopbackAsync(sslServerProtocol, sslClientProtocol)
-        return (
-            sslClientProtocol,
-            sslServerProtocol,
-            clientDataReceived,
-            serverDataReceived,
+        deferred = gatherResults(
+            [clientDataReceived, serverDataReceived], consumeErrors=True
         )
+        deferred.addCallback(lambda _: (sslClientProtocol, sslServerProtocol))
+        deferred.addErrback(lambda firstError: firstError.value.args[0])
+        return deferred
 
     def test_negotiationWithNoProtocols(self):
         """
         When factories support L{IProtocolNegotiationFactory} but don't
         advertise support for any protocols, no protocols are negotiated.
         """
-        (
-            client,
-            server,
-            clientDataReceived,
-            serverDataReceived,
-        ) = self.handshakeProtocols([], [])
+        deferred = self.handshakeProtocols([], [])
 
-        def checkNegotiatedProtocol(ignored):
+        def checkNegotiatedProtocol(result):
+            client, server = result
             self.assertEqual(client.negotiatedProtocol, None)
             self.assertEqual(server.negotiatedProtocol, None)
 
-        clientDataReceived.addCallback(lambda ignored: serverDataReceived)
-        serverDataReceived.addCallback(checkNegotiatedProtocol)
-
-        return clientDataReceived
+        return deferred.addCallback(checkNegotiatedProtocol)
 
     def test_negotiationWithProtocolOverlap(self):
         """
         When factories support L{IProtocolNegotiationFactory} and support
         overlapping protocols, the first protocol is negotiated.
         """
-        (
-            client,
-            server,
-            clientDataReceived,
-            serverDataReceived,
-        ) = self.handshakeProtocols([b"h2", b"http/1.1"], [b"h2", b"http/1.1"])
+        deferred = self.handshakeProtocols(
+            clientProtocols=[b"http/1.1", b"h2", b"client"],
+            serverProtocols=[b"h2", b"http/1.1", b"server"],
+        )
 
-        def checkNegotiatedProtocol(ignored):
+        def checkNegotiatedProtocol(result):
+            client, server = result
             self.assertEqual(client.negotiatedProtocol, b"h2")
             self.assertEqual(server.negotiatedProtocol, b"h2")
 
-        clientDataReceived.addCallback(lambda ignored: serverDataReceived)
-        serverDataReceived.addCallback(checkNegotiatedProtocol)
+        return deferred.addCallback(checkNegotiatedProtocol)
+
+    def test_negotiationNoOverlap(self):
+        """
+        When there is no overlap between client and server the connetion fails.
+
+        This test might fail on older OpenSSL version.
+        Prior to version 1.1.0 of OpenSSL, failing the ALPN negotiation does not
```

Comment:
It seems like this could be expressed as a skip?  The OpenSSL version in use should be available to us via either PyOpenSSL or cryptography.  I don't know if this is really a good idea or not because I don't know if there's a clear divide between OpenSSL where this test works and where it doesn't.  Is handshake failure because of ALPN failure a proper feature of OpenSSL now or just a random accident?

But a version-based skip seems a little bit better than just some words in the docstring.

Reply:
This tests is always green an all "supported" OpenSSL version.
Where supported mean OpenSSL version for which we run automated tests.

I left if without a skip, so that first time when we see it fails, we can review which version of OpenSSL is supported.
If that version of OpenSSL is supported and this test still fails, only then add the skip.

If the handshake doesn't fail, you end up with a "stalled"  / hangled / idle connection.

And I think that is important to know if the connection hangs and you can end up with a server with many such connections.

I can add a skip...but since so far this tests never fails... I decided that is best to left the comment for future debugging.

I guess that NPN is dead and people will use ALPN with 1.1.1 OpenSSL.

Also OpenSSL 1.1.0 is EOL  August 31, 2018 ... so maybe  just delete this comment without skipping this test as we will not support OpenSSL 1.1.0..

RHEL 7 has 1.0.2k-fips which was the LTS with EOL  December 31, 2019

RHEL 8 has 1.1.1c

The bottom line,  I don't know on which version I should  skip :) ... and I think that with current and future OpenSSL supported version this should never be skipped.

Reply:
> This tests is always green an all "supported" OpenSSL version.

I was only thinking of a skip for the known unsupported case.  But spelling it out explicitly like this, it doesn't seem worth it.  There are tons of unsupported versions of libraries that will make Twisted do wrong things and we  don't skip other tests for those cases.  So nevermind.

> Also OpenSSL 1.1.0 is EOL August 31, 2018 ... so maybe just delete this comment without skipping this test as we will not support OpenSSL 1.1.0..

This seems reasonable.  We don't go out of our way to make notes about other obsolete library versions.  OTOH, OpenSSL is complex and hard so the extra info isn't a bad thing.  Maybe make the comment more explicit about the fact that the information is left for future maintainers (actually, maybe a comment would be more appropriate for this information, compared to a docstring) and that the mentioned version of OpenSSL is explicitly not supported.